[Gradient Compression] Implement the original layerwise PowerSGD

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49451 [Gradient Compression] Warm-start of PowerSGD
* #49435 [Gradient Compression] Replace the assertions in PowerSGD comm hook by stream syncrhonization
* #49418 [Gradient Compression] Add error feedback to layerwise PowerSGD
* **#49417 [Gradient Compression] Implement the original layerwise PowerSGD**

The existing implementation applies PowerSGD to a batch of flattened tensors, which is a coarse-grained compression. This hook now is renamed as "batched_powerSGD_hook".

Now implement the original implementation in the paper, which applies PowerSGD to each per-parameter tensor. This is a layerwise fine-grained compression. Although this original implementation is slower, it is expected to achieve a higher accuracy, especially when the shapes of per-param tensors cannot be aligned.

Also add a test in distributed_test.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25511543](https://our.internmc.facebook.com/intern/diff/D25511543/)

Commit:
[Gradient Compression] Implement the original PowerSGD

The existing implementation applies PowerSGD to a batch of flatened tensors, which is a coarse-grained compression. This hook now is renamed as "batched_powerSGD_hook".

Now implement the original implementation in the paper, which applies PowerSGD to each per-parameter tensor. This is a layer-wise fine-grained compression. Although this original implementation is slower, it is expected to achieve a higher accuracy, especially when the shapes of per-param tensors cannot be aligned.

Also add a test in distributed_test.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25511543](https://our.internmc.facebook.com/intern/diff/D25511543/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Implement the original PowerSGD"


The existing implementation applies PowerSGD to a batch of flatened tensors, which is a coarse-grained compression. This hook now is renamed as "batched_powerSGD_hook".

Now implement the original implementation in the paper, which applies PowerSGD to each per-parameter tensor. This is a layer-wise fine-grained compression. Although this original implementation is slower, it is expected to achieve a higher accuracy, especially when the shapes of per-param tensors cannot be aligned.

Also add a test in distributed_test.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25511543](https://our.internmc.facebook.com/intern/diff/D25511543/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
     compression as follows:
+    1) Views the input flattened 1D gradient tensor as two groups of per-parameter tensors:
+    high-rank tensors and vector-like rank-1 tensors (for biases).
+    2) Handles rank-1 tensors by allreducing them without compression:
```
Does the paper do this shortcut for allreduce 1d tensors directly or is this something that we've added ourselves?

SciPioneer replied:
Same as the paper implementation. The input of the compression is a matrix, so it does not work on 1D tensors.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+    3) Handles high-rank tensors by PowerSGD compression:
+        3.1) For each high-rank tensor M, creates two low-rank tensors P and Q for decomposing M,
+        such that M = PQ^T, where Q is initialized from a standard normal distribution and orthogonalized;
+        3.2) Computes each P in Ps, which is equal to MQ;
```
How do we guarantee that Q = (Q^T)-1? I'm guessing this comes from the orthogonality of the matrix?

SciPioneer replied:
Exactly. This is basically the [definition of orthogonal matrix](https://en.wikipedia.org/wiki/Orthogonal_matrix).

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+    """
+    process_group = state.process_group
+    group_to_use = process_group if process_group is not None else dist.group.WORLD
+    world_size = (
```
nit: Can this just be `group_to_use.size()`?

SciPioneer replied:
Not sure. I copied this code from the implementations of other existing comm hooks.

Let me try to do the replacement for all the existing implementations in a separate PR and test the change thoroughly, if applicable.

SciPioneer replied:
Created another PR #49715 to do the replacement for all the existing comm hooks.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+
+    # This allreduce is only applied to rank-1 tensors,
+    # so it should have been kicked off before the above computation on the high-rank tensors to hide more communication costs.
+    # However, this somehow requires a separate future chain at this time.
```
Do we know why it requires a separate future chain? Can we file an issue to track this to ensure we don't miss the perf enhancement opportunity here?

SciPioneer replied:
The computation on rank-1 tensors and high-rank tensors are totally independent, ideally they can be parallelized.

However, the current future library does not have a native support -- it cannot easily support a fanout structure.

Added a TODO here.

SciPioneer replied:
Create #49716 to track this issue.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+    # This allreduce is only applied to rank-1 tensors,
+    # so it should have been kicked off before the above computation on the high-rank tensors to hide more communication costs.
+    # However, this somehow requires a separate future chain at this time.
+    allreduce_contiguous_rank1_tensors_fut = dist.all_reduce(
```
Can we skip this if there are no `rank1_tensors`, because then we're just allreducing an empty tensor?

SciPioneer replied:
It's very common to have rank-1 tensors in practice. For example, Y = aX + b, b (bias) is represented as a rank-1 tensor. I cannot imagine a real network that does not have any bias vector in practice.

On the other hand, even if there is no `rank1_tensors`, as shown in a unit test, the code can still pass.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+    ).get_future()
+
+    def unpack_rank1_tensors_and_allreduce_ps(fut):
+        rank1_tensors_memory = fut.value()[0].div_(world_size)
```
Is there a reason this isn't `fut.wait()`? Other calls seem to use `fut.wait()`.

SciPioneer replied:
Good question!

`wait()` is already called once in the precursor callback, in the return statement, and it should only be called once.

`value()` means just reading the value without blocking, and it's user's responsibility to ensure proper wait before retrieving the value.

There is a recent PR that changed the semantics of value and wait. See: https://github.com/pytorch/pytorch/pull/48505/files#r532577873

I think I should change more `wait()` into `value()` in the original PowerSGD implementation. Previously, these two functions are kind of equivalent on NCCL backend.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2769,6 +2770,52 @@ def test_DistributedDataParallel_non_default_stream(self):
+        def test_DistributedDataParallel_powerSGD_ddp_comm_hook(self):
+            stream = torch.cuda.Stream(self.rank)
+            rank = self.rank
+            with torch.cuda.stream(stream):
```
Is there something specific that we're aiming to test with using a separate stream?

SciPioneer replied:
I mainly followed the test case above this one, which tests allreduce.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2769,6 +2770,52 @@ def test_DistributedDataParallel_non_default_stream(self):
+                net = torch.nn.parallel.DistributedDataParallel(
+                    torch.nn.Linear(1, 5).to(rank), device_ids=[rank]
+                )
+                process_group = torch.distributed.new_group([0, 1])
```
Should we test with the process_group being the entire world as well?

SciPioneer replied:
I just followed some examples like `test_ddp_uneven_inputs_replicated_error` in the same file, and I didn't see it's necessary to testing both cases here.

I can change `torch.distributed.new_group([0, 1])` to `list(range(0, dist.get_world_size()))`, or just use `group, _, rank = self._init_global_test()`.

I plan to rewrite the unit test in a separate PR, as mentioned in some other comments.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2769,6 +2770,52 @@ def test_DistributedDataParallel_non_default_stream(self):
+                )
+                net.register_comm_hook(state=state, hook=powerSGD.powerSGD_hook)
+                # NOTE: batched_powerSGD_hook cannot pass the following test, because it has a lower accuracy.
+                for i in range(3):
```
This logic looks similar to `test_ddp_non_default_stream`, can you separate the duplicated logic into a helper function?

SciPioneer replied:
I realize that I probably should use a better example, not `test_DistributedDataParallel_non_default_stream`.

Let me update the unit test in a separate refactoring PR.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -2769,6 +2770,52 @@ def test_DistributedDataParallel_non_default_stream(self):
+                    world_size = int(os.environ["WORLD_SIZE"])
+                    avg.div_(world_size)
+                    expected_grad = sum(i for i in range(world_size)) / world_size
+                    self.assertEqual(
```
Hmm, I'm not sure if I quite understand this test. Due to the gradient compression computing a low-rank matrix approximation, allreduce should not necessarily give us back the average gradient right? Since we no longer have the exact gradient, but an approximation of it? Given that, I'm not sure why this test passes, can you elaborate on that?

SciPioneer replied:
Good question! Because the network in the test only has a single linear layer, so it can be considered as a Nx1 matrix, plus a bias vector. To compress this kind of matrix, there is no compression loss for PowerSGD.

Given a MxN matrix, if min(M, N) <= r, where r is the matrix approximation rank, there will be no loss for the compression.

rohan-varma replied:
Sounds good, would we also want to add a unittest that meaningfully compresses the grads?

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+
+    # Step II: Handle high-rank tensors.
+    # Allocate contiguous memory for Ps and Qs to allreduce compressed high-rank tensors efficiently.
+    high_rank_tensors = [
```
Nit: Wonder if checking if tensor not in rank1_tensors is faster (just does a pointer comparison I'm guessing), although it shouldn't make a significant difference either way.

SciPioneer replied:
`ndimension()` just retrieves a value from the metadata of tensor -- should be cheap too. Overall, I think it has a better readability than pointer comparison.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+        for p, q, tensor in zip(ps, qs, high_rank_tensors):
+            torch.matmul(p, q.t(), out=tensor)
+            assert not torch.any(torch.isnan(tensor))
+        return [input_tensor]
```
Why are we returning input_tensor here? Are we overwriting this tensor somewhere? Shouldn't we return the approximation PQ^T which is `tensor` in this case?

rohan-varma replied:
Ah I see, every `tensor` in high_rank_tensors is a view into `input_tensor` and modifying tensor also modifies the appropriate per-param tensor in `input_tensor`. If this is the case, could we clarify that is happening here (and why it's preferable to do this instead of directly return the high_rank_tensors here?

SciPioneer replied:
`high_rank_tensors` is a partial view of the input tensor (which is also the output here as the compression occurs in place). We need to return the input tensor, which is a combination of high-rank tensors and rank-1 tensors. This is the final output.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+            .wait()[0]
+        ]
+
+    def compute_qs(fut):
```
Nit: rename to `compute_and_allreduce_qs`?

SciPioneer replied:
I view allreduce as a part of computing qs -- basically it's computing local unsynced qs first and then global synced qs.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+        ]
+
+    def decompress(fut):
+        qs_memory = fut.wait()[0].div_(world_size)
```
where is `qs_memory` used in this function?

SciPioneer replied:
`qs_memory` is not explicitly used in this function, it is used by the operations on its view `qs` in this function.

The variable `qs` is a list of tensors, where the memory is allocated on top of `qs_memory`.  Allreducing `qs_memory` is equivalent to allreducing `qs`, as  a view of `qs_memory`.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+        q_idx += m * matrix_approximation_rank
+
+    # Initialize and then orthogonalize Qs.
+    with torch.random.fork_rng(devices=[]):
```
Would it be reasonable to dedupe other use cases of this forking in grad compression to a helper function?

SciPioneer replied:
Had the same feeling. However, in the first implementation, it has a loop in it: `for q in qs:`, between setting the manual seed and filling random values. It can be a bit tricky. Let me try to do it in a separate refactoring PR.

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -75,10 +75,193 @@ def powerSGD_hook(
+        n, m = tensor.shape
+        matrix_approximation_rank = min(n, m, state.matrix_approximation_rank)
+        ps.append(
+            ps_memory[p_idx : p_idx + n * matrix_approximation_rank].view(
```
Nit: seems like we are basically recomputing the size here to get the pointers to tensors with the right offsets, maybe it's faster to save the sizes we compute above and re-use them here.

SciPioneer replied:
Since the bucket can be rebuilt at some time, caching them won't be safe. It can also be a potential risk if we try to optimize the use cases that involve unused parameters in the future.

On the other hand, the computation here is actually very cheap, so I don't feel it's worth consuming to more memory for this purpose. The original paper implementation does not store this info neither.

rohan-varma reviewed (commented):
Thanks for working on this! I left a few comments/questions inline.

rohan-varma reviewed (approved):
Thanks for answering all my questions/comments! LGTM, we can refactor the tests in a follow up PR as you mentioned.

Commit:
Update on "[Gradient Compression] Implement the original layerwise PowerSGD"


The existing implementation applies PowerSGD to a batch of flattened tensors, which is a coarse-grained compression. This hook now is renamed as "batched_powerSGD_hook".

Now implement the original implementation in the paper, which applies PowerSGD to each per-parameter tensor. This is a layerwise fine-grained compression. Although this original implementation is slower, it is expected to achieve a higher accuracy, especially when the shapes of per-param tensors cannot be aligned.

Also add a test in distributed_test.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25511543](https://our.internmc.facebook.com/intern/diff/D25511543/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Implement the original layerwise PowerSGD"


The existing implementation applies PowerSGD to a batch of flattened tensors, which is a coarse-grained compression. This hook now is renamed as "batched_powerSGD_hook".

Now implement the original implementation in the paper, which applies PowerSGD to each per-parameter tensor. This is a layerwise fine-grained compression. Although this original implementation is slower, it is expected to achieve a higher accuracy, especially when the shapes of per-param tensors cannot be aligned.

Also add a test in distributed_test.py.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25511543](https://our.internmc.facebook.com/intern/diff/D25511543/)

[ghstack-poisoned]

mrshenli commented:
This breaks multi-gpu test on master. reverting

https://app.circleci.com/pipelines/github/pytorch/pytorch/253510/workflows/d97063d2-eeb6-470c-88c8-7a64553078fc/jobs/9759703

```
Dec 19 03:33:52   test_DistributedDataParallel_powerSGD_ddp_comm_hook (__main__.TestDistBackendWithFork) ... ERROR:root:Caught exception: 
Dec 19 03:33:52 Traceback (most recent call last):
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 283, in wrapper
Dec 19 03:33:52     fn()
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 97, in wrapper
Dec 19 03:33:52     return func(*args, **kwargs)
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/common_distributed.py", line 156, in wrapper
Dec 19 03:33:52     return func(*args, **kwargs)
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/testing/_internal/distributed/distributed_test.py", line 2851, in test_DistributedDataParallel_powerSGD_ddp_comm_hook
Dec 19 03:33:52     loss.backward()
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/tensor.py", line 233, in backward
Dec 19 03:33:52     torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)
Dec 19 03:33:52   File "/opt/conda/lib/python3.6/site-packages/torch/autograd/__init__.py", line 146, in backward
Dec 19 03:33:52     allow_unreachable=True, accumulate_grad=True)  # allow_unreachable flag
Dec 19 03:33:52 RuntimeError: AttributeError: 'object' object has no attribute 'size'
Dec 19 03:33:52 
Dec 19 03:33:52 At:
Dec 19 03:33:52   /opt/conda/lib/python3.6/site-packages/torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py(118): powerSGD_hook
Dec 19 03:33:52 
Dec 19 03:33:52 exiting process with exit code: 10
Dec 19 03:33:53 Process 2 terminated with exit code 10, terminating remaining processes.
Dec 19 03:33:53 ERROR (2.849s)
```
